neqo-transport/src/crypto.rs

```diff
@@ -1358,6 +1360,80 @@ pub enum CryptoStreams {
     },
 }
 
+/// [Fisher–Yates shuffle](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#Modern_method)
+/// Modified to make sure no element stays in place if `a` has more than one element.
+///
+/// TODO: We could use <https://docs.rs/rand/latest/rand/seq/trait.SliceRandom.html#tymethod.shuffle>
+/// instead, but I assume we want to only use the NSS RNG in neqo?
+fn shuffle<T>(a: &mut [T]) {
+    // To shuffle an array a of n elements (indices 0..n-1)
+    let n: usize = a.len();
+    if n < 2 {
+        return;
+    }
+    for i in 0..(n - 1) {
+        // j ← random integer such that i ≤ j < n
+        let j = loop {
+            let j = (random::<8>()[0] as usize) % (n - i) + i;
+            if j != i {
+                break j;
+            }
+        };
+        // Exchange a[i] and a[j]
+        a.swap(i, j);
+    }
+}
+
+/// Find the start and end indices of all sequences of `len` or more graphic ASCII bytes in `data`.
+fn ascii_sequences(data: &[u8], len: usize) -> Vec<(usize, usize)> {
+    let mut sequences = vec![];
+    let mut start = None;
+    for (i, &b) in data.iter().enumerate() {
+        if b.is_ascii_graphic() {
+            if start.is_none() {
+                start = Some(i);
+            }
+        } else if let Some(s) = start {
+            if i - s >= len {
+                sequences.push((s, i));
+            }
+            start = None;
+        }
+    }
+    if let Some(s) = start {
+        if data.len() - s >= len {
+            sequences.push((s, data.len()));
+        }
+    }
+    sequences
+}
+
+/// Look for ranges of `N` or more bytes of graphical ASCII data in `data`. Create at least one
+/// split point at `N/2` into the range for each range, multiple ones each `N` bytes afterwards if
+/// the range is long enough. Create data chunks based on those split points. Shuffle the chunks and
+/// return them.
+fn reorder_chunks(data: &[u8]) -> Vec<(u64, &[u8])> {
+    const N: usize = 3;
+    let mut splits = vec![];
+    // For each sequence, split it into chunks of `N` bytes.
+    for (mut start, end) in ascii_sequences(data, N) {
+        while start + N <= end {
+            splits.push(start + N / 2);
+            start += N;
+        }
```

Comment:
Splitting a string multiple times is probably unnecessary.

This might be easier as:

```rust
let mut used = 0;
let ranges = ascii_sequences(data, N);
let mut chunks = Vec::with_capacity(ranges.len() + 1);
for (start, end) = ranges {
  debug_assert!(start > used);
  let (left, right) = data.split_at(start + N / 2 - used);
  chunks.push((u64::try_from(used).unwrap(), left));
  used += left;
  data = right;
}
shuffle(&mut chunks);
chunks
```

Reply:
This fails a bunch of tests. But also, why not split longer ASCII ranges more than once? See above for performance.